src/workflow/workflow_tasks/concurrency_manager.rs

```diff
@@ -0,0 +1,284 @@
+use crate::{
+    errors::WorkflowUpdateError,
+    protos::{
+        coresdk::workflow_activation::WfActivation, temporal::api::common::v1::WorkflowExecution,
+    },
+    protosext::ValidPollWFTQResponse,
+    workflow::{
+        workflow_tasks::{OutstandingActivation, OutstandingTask},
+        HistoryUpdate, Result, WorkflowError, WorkflowManager,
+    },
+};
+use futures::future::{BoxFuture, FutureExt};
+use parking_lot::{RwLock, RwLockReadGuard, RwLockWriteGuard};
+use std::{
+    collections::HashMap,
+    fmt::Debug,
+    ops::{Deref, DerefMut},
+};
+use tokio::sync::Mutex;
+
+/// Provides a thread-safe way to access workflow machines for specific workflow runs
+pub(crate) struct WorkflowConcurrencyManager {
+    /// Maps run id -> data about and machines for that run
+    machines: RwLock<HashMap<String, ManagedRun>>,
+}
+
+struct ManagedRun {
+    wfm: Mutex<Option<WorkflowManager>>,
```

Comment:
Why do we need the mutex here and only here?

Reply:
There's another comment about this, but basically it allows the read lock to be held instead of the write lock on the overall map. Since most work happens in here, it's more efficient (theoretically).

I don't want to bother tweaking this until I have a change to move this all inside of a worker per the other comment. Then will be a good time to simplify.